File: 63-unique-paths-ii/63-unique-paths-ii.cpp
class Solution {
private:
    int f(int m, int n, vector<vector<int>>& obstacleGrid, vector<vector<int>> &dp) {
        if(m < 0 or n < 0 or obstacleGrid[m][n] == 1) return 0;
        if(m == 0 and n == 0) return 1;
        if(dp[m][n] != -1) return dp[m][n];
        int left = 0, up = 0;
        if(m > 0) left = f(m - 1, n, obstacleGrid, dp);
        if(n > 0) up = f(m, n - 1, obstacleGrid, dp);
        return dp[m][n] = left + up;
    }
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size(), n = obstacleGrid[0].size();
        vector<vector<int>> dp(m, vector<int>(n, -1));
        return f(m - 1, n - 1, obstacleGrid, dp);
    }
};

File: 63-unique-paths-ii/63-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        prevRow = [0] * (n)
        for i in range (m):
            curRow = [0] * (n)
            for j in range (n):
                if obstacleGrid[i][j] == 1:
                    curRow[j] = 0
                    continue
                if i == 0 and j == 0: 
                    curRow[0] = 1
                    continue
                down, right = 0, 0
                if j > 0: right = curRow[j - 1]
                if i > 0: down = prevRow[j]
                curRow[j] = right + down 
            prevRow = curRow
        return prevRow[n - 1]